File: Q3_PG_Driver_Interface/driver.py
import psycopg2
import json
from decimal import Decimal

class DecimalEncoder(json.JSONEncoder):
    def default(self, o):
        if isinstance(o, Decimal):
            return int(o)
        return super().default(o)

# Connect to the PostgreSQL database
conn = psycopg2.connect(
    host="your_host",
    database="your_database",
    user="your_user",
    password="your_password"
)

# Create a cursor object to execute SQL queries
cur = conn.cursor()

# Execute a SELECT query to retrieve data from the table
cur.execute("SELECT * FROM public.user_table")

# Fetch all rows returned by the query
rows = cur.fetchall()

# Define a list to store the rows as dictionaries
result = []

# Convert each row to a dictionary and append to the result list
for row in rows:
    user = {
        "user_id": row[0],
        "name": row[1],
        "age": row[2]
    }
    if row[3] is not None:
        user["phone"] = row[3]

    result.append(user)

# Create the final JSON object
json_output = {
    "status_code": 200,
    "data": result
}

# Convert the JSON object to a string
json_string = json.dumps(json_output, cls=DecimalEncoder)

# Print the JSON string
print(json_string)

# Close the cursor and connection
cur.close()
conn.close()

File: Q3_PG_Driver_Interface/Node/driver.js
// Import the required libraries
const { Pool } = require('pg');

// Create a function to fetch data from the user_table in JSON format
async function fetchUserData() {
  // Create a new instance of the Pool class with your PostgreSQL connection details
  const pool = new Pool({
    user: 'your_user',
    host: 'your_host',
    database: 'your_database',
    password: 'your_password',
    port: 'your_port',
  });

  // Connect to the PostgreSQL database
  const client = await pool.connect();

  try {
    // Execute the SQL query to fetch user data from the user_table
    const query = 'SELECT user_id, name, age, phone FROM public.user_table ORDER BY user_id ASC';
    const result = await client.query(query);

    // Remove null values from the result rows
    const filteredRows = result.rows.map(row => {
      const filteredRow = {};
      for (const key in row) {
        if (row[key] !== null) {
          filteredRow[key] = row[key];
        }
      }
      return filteredRow;
    });

    // Format the query result as a JSON string
    const jsonData = JSON.stringify({
      status_code: 200,
      data: filteredRows,
    });

    // Return the JSON string
    return jsonData;
  } finally {
    // Release the client and the connection pool
    client.release();
    pool.end();
  }
}

// Call the fetchUserData function and log the result
fetchUserData()
  .then((jsonData) => console.log(jsonData))
  .catch((error) => console.error('Error:', error));
